### PR TITLE
SMHE-766 Key change g-meter

### PR DESCRIPTION
**Pull request by stefanermens**

### CODE REVIEW

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/SetEncryptionKeyExchangeOnGMeterRequestFactory.java [hunk 1]
```diff
@@ -26,6 +28,19 @@ public static SetEncryptionKeyExchangeOnGMeterRequest fromParameterMap(
...
+    final SetEncryptionKeyExchangeOnGMeterRequestData data =
+        new SetEncryptionKeyExchangeOnGMeterRequestData();
+
+    data.setMbusDeviceIdentification(
+        requestParameters.get(PlatformSmartmeteringKeys.KEY_DEVICE_IDENTIFICATION));
+    data.setSecretType(
+        SecretType.valueOf(requestParameters.get(PlatformSmartmeteringKeys.SECRET_TYPE)));
+    data.setCloseOpticalPort(
+        Boolean.valueOf(
+            requestParameters.getOrDefault(PlatformSmartmeteringKeys.CLOSE_OPTICAL_PORT, "false")));
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/SetEncryptionKeyExchangeOnGMeterRequestFactory.java:40]: Hoort "false" niet in de PlatformSmartmeteringDefaults thuis?
 _PlatformSmartmeteringDefaults.CLOSE_OPTICAL_PORT_

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java [hunk 2]
```diff
@@ -503,28 +512,31 @@ public void aValidMbusUserKeyIsStored(final Map<String, String> settings) {
...
     }
 
     assertThat(numberOfMbusDefaultKeys).as("Number of M-Bus Default keys stored").isEqualTo(1);
-    assertThat(numberOfMbusUserKeys > 0).as("At least one M-Bus User key must be stored").isTrue();
-    assertThat(numberOfValidMbusUserKeys).as("Number of valid M-Bus User keys stored").isEqualTo(1);
+    assertThat(numberOfMbusKeys >= storedKeyCount)
+        .as(
+            "At least " + storedKeyCount + " M-Bus key of the specified type must be stored",
+            storedKeyCount)
+        .isTrue();
```
Comment (harrymiddelburgetall) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java:534]: assertThat(numberOfMbusKeys).**isGreaterThanOrEqualTo**(storedKeyCount).as(...)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/SetKeyOnGMeterCommandExecutor.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+import org.opensmartgridplatform.shared.exceptionhandling.EncrypterException;
+import org.opensmartgridplatform.shared.infra.jms.MessageMetadata;
+import org.opensmartgridplatform.ws.schema.core.secret.management.SecretType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component()
+public class SetKeyOnGMeterCommandExecutor
```
Comment (harrymiddelburgetall) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/SetKeyOnGMeterCommandExecutor.java:48]: It is a bit confusing to introduce a new CommandExecutor with a name that is more appropriate to server different types of keys but to introduce a new RequestDto that still follows the old naming.

File: osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/SetEncryptionKeyExchangeOnGMeterRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.dto.valueobjects.smartmetering;
+
+public class SetEncryptionKeyExchangeOnGMeterRequestDto implements ActionRequestDto {
```
Comment (harrymiddelburgetall) [inline osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/SetEncryptionKeyExchangeOnGMeterRequestDto.java:12]: Why not naming this newly introduced requestDto **SetKeyOnGMeterRequestDto** which is more appropriate as you gave a new name to the newly implemented command executor

**Review by harrymiddelburgetall — APPROVED** [review]
Nice work on this complex matter. Tests with real meters will prove the correctness of the code as I can't.  As you mentioned before some renaming is needed but newly introduced classes could have a name which is more in line straight away I suppose. My other comments are minors.

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/SetEncryptionKeyExchangeOnGMeterRequestFactory.java (hunk 1 above)
Comment (natasjaalliander) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/SetEncryptionKeyExchangeOnGMeterRequestFactory.java]: yes, PlatformSmartmeteringDefaults.CLOSE_OPTICAL_PORT is added.

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java (hunk 2 above)
Comment (natasjaalliander) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/database/device/DlmsDeviceSteps.java]: assert is updated.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/SetKeyOnGMeterCommandExecutor.java (hunk 3 above)
Comment (natasjaalliander) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/SetKeyOnGMeterCommandExecutor.java:48]: right, name of the requestDto is updated to SetKeyOnGMeterRequestDto, in line with the SetKeyOnGMeterCommandExecutor

File: osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/SetEncryptionKeyExchangeOnGMeterRequestDto.java (hunk 4 above)
Comment (natasjaalliander) [reply osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/SetEncryptionKeyExchangeOnGMeterRequestDto.java]: right, name of the requestDto is updated to SetKeyOnGMeterRequestDto, in line with the SetKeyOnGMeterCommandExecutor

**Review by abolder — APPROVED** [review]
Opmerkingen zijn aangepast dus wat mij betreft akkoord

**Pull request merged**